packages/lib/src/vis/scatter/ScatterPoints.tsx

```diff
@@ -71,28 +72,46 @@ function ScatterPoints(props: Props) {
     [onPointerOut]
   );
 
-  const [dataGeometry] = useState(() => new BufferGeometry());
+  const { length } = data;
+  const { abscissaScale, ordinateScale } = useVisCanvasContext();
   const invalidate = useThree((state) => state.invalidate);
 
-  const dataToColorScale = useDataToColorScale(
+  const indexToPosition = useIndexToPosition(
+    abscissas,
+    abscissaScale,
+    ordinates,
+    ordinateScale
+  );
+
+  const valueToColor = useValueToColor(
     scaleType,
     domain,
     colorMap,
     invertColorMap
   );
 
-  const { position, color } = useBufferAttributes(
-    abscissas,
-    ordinates,
-    data,
-    dataToColorScale
-  );
+  const dataGeometry = useMemo(() => {
+    const geometry = new BufferGeometry();
+    geometry.setAttribute('position', createBufferAttr(length, 3));
+    geometry.setAttribute('color', createBufferAttr(length, 3));
+    return geometry;
+  }, [length]);
 
   useLayoutEffect(() => {
```

Comment:
I'm now looping through the `data` array inside the `useLayoutEffect` instead of in a `useMemo` beforehand. This allows me to mutate the buffers instead of recreating instances.